checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -362,57 +398,128 @@ private static int getGroupCount(Pattern p) {
     }
 
     /**
-     * Return the non-null groups in the argument.
+     * Returns a list of groups other than 0, that are guaranteed to be non-null given that the
+     * regular expression matches the String.
      *
-     * @param p pattern to be analysed
+     * @param regexp pattern to be analysed
+     * @param n number of capturing groups in the pattern
      * @return the groups that are non-null in the argument
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
         List<Integer> nonNullGroups = new ArrayList<>();
```

Comment:
As a minor point, a BitSet is more efficient than a list.

Reply:
Do you mean a HashSet or a BitSet? Isn't BitSet a binary array with support for operations like shifts, basically a number with many bits? Or do you mean I should set the bits where the groups are non-null and then convert it to an array by traversing the bitset?

Reply:
This list is encoding, for each integer, whether the group is non-null or nullable.  That is, it is encoding a boolean value for each integer.

That concept can be implemented in a number of ways:
 * by listing all the integers that map to true.  Determine the boolean value for `i` as `list.contains(i)`, which has to search the entire array.  Lists have a large run-time representation, and adding a value may require memory allocation.
 * as an array of booleans.  Determine the boolean value for `i` as `array[i]`, which is constant-time.
 * as a BitSet, which has all the advantages of an array of booleans but a much smaller run-time representation.

I was suggesting the latter.  This is a minor point, however.

Reply:
For some reason, my BitSet implementation is exceeding the memory limits. [Here](https://dev.azure.com/utkarshrai/Checker%20Framework/_build/results?buildId=107&view=results). The ArrayList version of the same implementation passes plume-lib tests locally (the current CI jobs will pass too). Since this is a minor point, can we skip this for the time being?

Reply:
Yes, that is fine.  It is a minor point, after all.